samples/example_vptr/example_vptr.cpp

```diff
@@ -0,0 +1,115 @@
+
+#include <CL/sycl.hpp>
+
+#include <iostream>
+
+#include "virtual_ptr.hpp"
+
+using namespace cl::sycl;
+
+const size_t N = 100;
+const size_t M = 150;
+
+/* This sample allocates three device-only matrices, using the virtual pointer
+ * and SYCLmalloc. It initalises the first two in parallel on the device. After
+ * that, it adds them together, storing the result in the third matrix. It then
+ * verifies the result on the host by using:
+ *  - pointer arithmetic on the virtual pointer to index the matrix
+ *  - host accessor to gain access to the data. */
+int main() {
+  {
+    queue myQueue;
+    cl::sycl::codeplay::PointerMapper pMap;
+
+    /* Allocate the matrices using SYCLmalloc. a, b and c are virtual pointers,
+     * pointing to device buffers.
+     */
+    float* a = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* b = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+    float* c = static_cast<float*>(SYCLmalloc(N * M * sizeof(float), pMap));
+
+    /* This kernel will initialise the buffer pointed to by a. The accessor "A"
+     * has write access. We retrieve it directly from the PointerMapper, using
+     * the virtual pointer.*/
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::discard_write,
+                               access::target::global_buffer, float>(a, cgh);
+      cgh.parallel_for<class init_a>(
+          range<1>{N * M}, [=](item<1> index) { A[index] = index[0] * 2; });
+    });
+
+    /* Similarly, this kernel will initialise the buffer pointed to by b. */
+    myQueue.submit([&](handler& cgh) {
+      auto B = pMap.get_access<access::mode::discard_write,
+                               access::target::global_buffer, float>(b, cgh);
+      cgh.parallel_for<class init_b>(
+          range<1>{N * M}, [=](item<1> index) { B[index] = index[0] * 2014; });
+    });
+
+    /* This kernel will perform the computation C = A + B. */
+    myQueue.submit([&](handler& cgh) {
+      auto A = pMap.get_access<access::mode::read,
+                               access::target::global_buffer, float>(a, cgh);
+      auto B = pMap.get_access<access::mode::read,
+                               access::target::global_buffer, float>(b, cgh);
+      auto C = pMap.get_access<access::mode::discard_write,
+                               access::target::global_buffer, float>(c, cgh);
+      cgh.parallel_for<class matrix_add>(range<1>{N * M}, [=](item<1> index) {
+        C[index] = A[index] + B[index];
+      });
+    });
+
+    /* On the host, the result stored in the buffer of virtual pointer "c" are
+     * checked. The matrix is accessed row by row, using pointer arithmetics on
+     * the virtual pointer. */
+    float* c_row = c;
+    for (size_t i = 0; i < N; i++) {
+      /* Get the number of elements by which the row is offset. */
+      auto row_offset = pMap.get_offset(c_row) / sizeof(float);
```

Comment:
I guess we should provide an interface to return the index directly for the basic type, to avoid the division.

Reply:
Good idea. Done. 444504bbbb3701a118247ecb499ad5843ef92d10